### PR TITLE
TextEdit cleanup and refactoring

### LINKED ISSUE #31739: Refactor TextEdit into two classes (text and code focused)
`TextEdit` is a generic control exposed to games and the scripting API for text input from the user. It's also the base control used in the various code editors of the engine (GDScript editor, `TextFile` editor for arbitrary text files, shader editor). Due to that, many features have been added to its API for the sole purpose of editor tools, making the class harder to maintain and with a sizeable share on unexposed, editor-specific APIs.

IMO we should avoid adding too many editor-specific APIs to base controls, and we should instead of editor-specific controls that inherit and extend the generic ones. Those may or may not be exposed to the runtime depending on whether they could be useful for e.g. plugins or in-game tools, but they should be separate and well identified controls.

As such I would suggest to split `TextEdit` into two controls:
- A slimmed down `TextEdit` that has a general purpose API, to write/copy/cut/paste text.
- A code-focused `CodeEdit`, that inherits from `TextEdit` but adds code-specific features such as indentation management, line numbers, code folding, etc.

Possibly, there could even be a third control which would be editor-only, while the above two would be exposed in the core API:
- An extended `EditorCodeEdit`, that inherits `CodeEdit` but adds more features specific to Godot's code editors, such as inline color markers, breakpoint gutters, etc. If we feel it's overkill, we can keep those in `CodeEdit` though, but it's worth discussing.

If we agree on the need to refactor this, we should spend some time designing the APIs to make sure that they can be easily extended for editor needs without having to hack `TextEdit`.
A recent example: in #28561 I added a feature to convert clipboard indentation on paste, but I can't have it overwrite <kbd>Ctrl</kbd>+<kbd>V</kbd> because that's hardcoded in `TextEdit`. The new API should allow overriding `TextEdit`'s input management in `CodeEdit` and `EditorCodeEdit`.

### PR DESCRIPTION
**Pull request by Paulb23**
Continuation of #31739

Builds on #50122

---

This PR seems larger then it is, mainly moving around code and renames, rather then any changes.

The main aim of this PR is to cleanup after extracting features into `CodeEdit` and to prepare the API going forward.  Will do `EditorSettings` as a separate PR.

Given `TextEdit` has organically grown rather large it became a bit of mess, so I took the time here to reorganise the methods; hopefully will made it easier to browse.

With that said, the three main API changes are:

- Block caret no longer exists and is now an `enum CaretType`, currently with `Line` and `Block` as the options. The methods have been renamed appropriately:
    - `cursor_set_block_mode` --> `set_caret_type`
    - `cursor_is_block_mode` --> `get_caret_type`
- Line wrapping has now been changed from enable / disable to  an `enum LineWrappingMode`, with `None` and `Boundary` meaning the control boundary as it currently works. Similarly the methods have been renamed:
    - `set_wrap_enabled` --> `set_line_wrapping_mode`
    - `is_wrap_enabled`  --> `get_line_wrapping_mode`
- Cut, copy and paste are now overrideable with:
    - `_cut`
    - `_copy`
    - `_paste  `

Otherwise bedsides exposing many new methods, the following renames have taken place:

Caret:
- `cursor_get_blink_enabled` --> `set_caret_blink_enabled`
- `cursor_set_blink_enabled` --> `is_caret_blink_enabled`
- `cursor_set_blink_speed` --> `set_caret_blink_speed`
- `cursor_get_blink_speed` --> `get_caret_blink_speed`
- `set_right_click_moves_caret` --> `set_move_caret_on_right_click_enabled`
- `is_right_click_moving_caret` --> `is_move_caret_on_right_click_enabled`
- `set_mid_grapheme_caret_enabled` --> `set_caret_mid_grapheme_enabled`
- `get_mid_grapheme_caret_enabled` --> `is_caret_mid_grapheme_enabled`
- `cursor_set_line` --> `set_caret_line`
- `cursor_get_line` --> `get_caret_line`
- `cursor_set_column` --> `set_caret_column`
- `cursor_get_column` --> `get_caret_column`
- `adjust_viewport_to_cursor` --> `adjust_viewport_to_caret`
- `insert_text_at_cursor` --> `insert_text_at_caret`
- `get_word_under_cursor` --> `get_word_under_caret`

Selection:
- `is_selection_active` --> `has_selection`
- `get_selection_text` --> `get_selected_text`

Insert Mode:
- `set_insert_mode` --> `set_overtype_mode_enabled`
- `is_insert_mode` --> `is_overtype_mode_enabled`

Read only:
- `set_readonly` --> `set_editable`
- `is_readonly` --> `is_editable`

Signals:
- `cursor_changed` --> `caret_changed`

Properties:
- `caret_block_mode` --> `caret_type`
- `caret_enabled` --> `wrap_mode`
- `caret_moving_by_right_click` --> `caret_move_on_right_click`

---

Supersedes #47548

Apart from the couple features that need a little more work, after this I'm happy to close #31739

### CODE REVIEW

File: editor/code_editor.cpp [hunk 1]
```diff
@@ -141,26 +141,23 @@ void FindReplaceBar::_unhandled_input(const Ref<InputEvent> &p_event) {
 }
 
 bool FindReplaceBar::_search(uint32_t p_flags, int p_from_line, int p_from_col) {
-	int line, col;
 	String text = get_search_text();
+	Vector2 pos = text_editor->search(text, p_flags, p_from_line, p_from_col);
```
Comment (bruvzg) [inline editor/code_editor.cpp:145]: `Vector2i` is probably more suitable.

File: scene/gui/text_edit.h [hunk 2]
```diff
@@ -63,6 +57,19 @@ class TextEdit : public Control {
 		SELECTION_MODE_LINE
 	};
 
+	/* Line Wrapping.*/
+	enum LineWrappingMode {
+		LINE_WRAPPING_NONE,
+		LINE_WRAPPING_BOUNDRY
```
Comment (bitsawer) [inline scene/gui/text_edit.h:63]: I wonder if this is a typo and it should be "LINE_WRAPPING_BOUND**A**RY"?

File: doc/classes/CodeEdit.xml [hunk 3]
```diff
@@ -645,7 +645,7 @@
 			The tint of text outline of the [CodeEdit].
 		</theme_item>
 		<theme_item name="font_readonly_color" data_type="color" type="Color" default="Color(0.88, 0.88, 0.88, 0.5)">
-			Sets the font [Color] when [member TextEdit.readonly] is enabled.
+			Sets the font [Color] when [member TextEdit.editable] is enabled.
```
Comment (KoBeWi) [inline doc/classes/CodeEdit.xml:648]: Shouldn't it be opposite (disabled)? Same below.

File: doc/classes/TextEdit.xml [hunk 4]
```diff
@@ -13,93 +13,99 @@
...
 			<return type="void" />
 			<description>
-				A virtual method that is called whenever backspace is triggered.
+				Override this method to define what happens when the user presses the backspace key.
+			</description>
+		</method>
+		<method name="_copy" qualifiers="virtual">
+			<return type="void" />
+			<description>
+				Override this method to define what happens when the user performs a copy.
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:22]:
```suggestion
				Override this method to define what happens when the user performs a copy operation.
```
Same for cut/paste.

File: doc/classes/TextEdit.xml [hunk 5]
```diff
@@ -13,93 +13,99 @@
...
+				Override this method to define what happens when the user performs a paste.
 			</description>
 		</method>
-		<method name="backspace">
+		<method name="add_gutter">
 			<return type="void" />
+			<argument index="0" name="at" type="int" default="-1" />
 			<description>
-				Causes the [TextEdit] to perform a backspace.
+				Register a new gutter to this [TextEdit]. [code]at[/code] to have a specific gutter order.
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:48]:
```suggestion
				Register a new gutter to this [TextEdit]. Use [code]at[/code] to have a specific gutter order.
```
Also it should mention what the -1 default means.

File: doc/classes/TextEdit.xml [hunk 6]
```diff
@@ -13,93 +13,99 @@
...
-				Removes all OpenType features.
+				Called when the user presses the backspace key. Can be overriden with [method _backspace].
 			</description>
 		</method>
-		<method name="clear_undo_history">
+		<method name="begin_complex_operation">
 			<return type="void" />
 			<description>
-				Clears the undo history.
+				Starts a multipart edit, all edits will be treated as one action until [method end_complex_operation] is called.
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:66]:
```suggestion
				Starts a multipart edit. All edits will be treated as one action until [method end_complex_operation] is called.
```

File: doc/classes/TextEdit.xml [hunk 7]
```diff
@@ -13,93 +13,99 @@
...
+				Centers the viewport on the line the editing cursor is at. This also resets the [member scroll_horizontal] value to [code]0[/code].
 			</description>
 		</method>
-		<method name="cursor_get_column" qualifiers="const">
-			<return type="int" />
+		<method name="clear">
+			<return type="void" />
 			<description>
-				Returns the column the editing cursor is at.
+				Performs a full reset of [TextEdit].
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:78]:
```suggestion
				Performs a full reset of [TextEdit], including undo history.
```
Probably worth to mention it.

File: doc/classes/TextEdit.xml [hunk 8]
```diff
@@ -13,93 +13,99 @@
...
-				Centers the viewport on the line the editing cursor is at. This also resets the [member scroll_horizontal] value to [code]0[/code].
+				Adjust the viewport so the caret is visible.
 			</description>
 		</method>
-		<method name="clear_opentype_features">
+		<method name="backspace">
 			<return type="void" />
 			<description>
-				Removes all OpenType features.
+				Called when the user presses the backspace key. Can be overriden with [method _backspace].
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:60]:
```suggestion
				Called when the user presses the backspace key. Can be overridden with [method _backspace].
```

File: doc/classes/TextEdit.xml [hunk 9]
```diff
@@ -13,93 +13,99 @@
...
 			<return type="void" />
-			<argument index="0" name="line" type="int" />
-			<argument index="1" name="adjust_viewport" type="bool" default="true" />
-			<argument index="2" name="can_be_hidden" type="bool" default="true" />
-			<argument index="3" name="wrap_index" type="int" default="0" />
 			<description>
-				Moves the cursor at the specified [code]line[/code] index.
-				If [code]adjust_viewport[/code] is set to [code]true[/code], the viewport will center at the cursor position after the move occurs.
-				If [code]can_be_hidden[/code] is set to [code]true[/code], the specified [code]line[/code] can be hidden.
+				Copy's the current text selection. Can be overriden with [method _copy].
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:96]:
```suggestion
				Copies the current text selection. Can be overridden with [method _copy].
```

File: doc/classes/TextEdit.xml [hunk 10]
```diff
@@ -13,93 +13,99 @@
...
 			</description>
 		</method>
 		<method name="cut">
 			<return type="void" />
 			<description>
-				Cut's the current selection.
+				Cut's the current selection. Can be overriden with [method _cut].
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:102]:
```suggestion
				Cut's the current selection. Can be overridden with [method _cut].
```

File: doc/classes/TextEdit.xml [hunk 11]
```diff
@@ -13,93 +13,99 @@
...
 			</description>
 		</method>
 		<method name="delete_selection">
 			<return type="void" />
 			<description>
+				Deletes the selected test.
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:108]:
```suggestion
				Deletes the selected text.
```

File: doc/classes/TextEdit.xml [hunk 12]
```diff
@@ -108,46 +114,99 @@
 				Deselects the current selection.
 			</description>
 		</method>
+		<method name="end_complex_operation">
+			<return type="void" />
+			<description>
+				Ends a multipart edit, started with [method begin_complex_operation].
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:120]: Could also mention what happens if no complex operation was started.

File: doc/classes/TextEdit.xml [hunk 13]
```diff
@@ -13,93 +13,99 @@
...
-				Clears the undo history.
+				Starts a multipart edit, all edits will be treated as one action until [method end_complex_operation] is called.
 			</description>
 		</method>
-		<method name="copy">
+		<method name="center_viewport_to_caret">
 			<return type="void" />
 			<description>
-				Copy's the current text selection.
+				Centers the viewport on the line the editing cursor is at. This also resets the [member scroll_horizontal] value to [code]0[/code].
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:72]:
```suggestion
				Centers the viewport on the line the editing caret is at. This also resets the [member scroll_horizontal] value to [code]0[/code].
```

File: doc/classes/TextEdit.xml [hunk 14]
```diff
@@ -108,46 +114,99 @@
...
+			<return type="int" />
+			<description>
+				Returns the wrap index the editing caret is on.
 			</description>
 		</method>
 		<method name="get_first_non_whitespace_column" qualifiers="const">
 			<return type="int" />
 			<argument index="0" name="line" type="int" />
 			<description>
+				Gets the column contain a non-whitespace character.
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:151]:
```suggestion
				Gets the first column containing a non-whitespace character.
```

File: doc/classes/TextEdit.xml [hunk 15]
```diff
@@ -157,13 +216,20 @@
...
 			<argument index="0" name="line" type="int" />
 			<description>
 				Returns the current background color of the line. [code]Color(0, 0, 0, 0)[/code] is returned if no color is set.
 			</description>
 		</method>
+		<method name="get_line_column_at_pos" qualifiers="const">
+			<return type="Vector2i" />
+			<argument index="0" name="position" type="Vector2i" />
+			<description>
+				Gets the line and column at the given position. In the return vector, [code]x[/code] is the column, [code]y[/code] is the line.
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:230]:
```suggestion
				Gets the line and column at the given position. In the returned vector, [code]x[/code] is the column, [code]y[/code] is the line.
```

File: doc/classes/TextEdit.xml [hunk 16]
```diff
@@ -175,27 +241,73 @@
...
+			<description>
+				Gets the hight of a largest line.
+			</description>
+		</method>
+		<method name="get_line_width" qualifiers="const">
+			<return type="int" />
+			<argument index="0" name="line" type="int" />
+			<argument index="1" name="wrap_index" type="int" default="-1" />
+			<description>
+				Gets the width in pixels of the [code]line[/code] [code]wrap_index[/code].
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:282]:
```suggestion
				Gets the width in pixels of the [code]line[/code] at the [code]wrap_index[/code].
```
idk, something is missing here

File: doc/classes/TextEdit.xml [hunk 17]
```diff
@@ -175,27 +241,73 @@
...
+			</description>
+		</method>
+		<method name="get_line_wrap_count" qualifiers="const">
+			<return type="int" />
+			<argument index="0" name="line" type="int" />
+			<description>
+				Returns the number of the time given line is wrapped.
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:289]:
```suggestion
				Returns the number of times the given line is wrapped.
```

File: doc/classes/TextEdit.xml [hunk 18]
```diff
@@ -175,27 +241,73 @@
...
+			<argument index="1" name="column" type="int" />
+			<description>
+				Returns the wrap index of the given line column.
+			</description>
+		</method>
+		<method name="get_line_wrapped_text" qualifiers="const">
+			<return type="PackedStringArray" />
+			<argument index="0" name="line" type="int" />
+			<description>
+				Returns an array of [String] representing each wrapped index.
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:304]:
```suggestion
				Returns an array of [String]s representing each wrapped index.
```

File: doc/classes/TextEdit.xml [hunk 19]
```diff
@@ -175,27 +241,73 @@
...
+			</description>
+		</method>
+		<method name="get_local_mouse_pos" qualifiers="const">
+			<return type="Vector2" />
+			<description>
+				Gets the local mouse position adjust for the text direction.
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:310]:
```suggestion
				Gets the local mouse position adjusted for the text direction.
```

File: doc/classes/TextEdit.xml [hunk 20]
```diff
@@ -204,16 +316,67 @@
...
+				Gets the total amount of lines that can be draw on the minimap.
+			</description>
+		</method>
+		<method name="get_next_visible_line_index_offset_from" qualifiers="const">
+			<return type="Vector2i" />
+			<argument index="0" name="line" type="int" />
+			<argument index="1" name="wrap_index" type="int" />
+			<argument index="2" name="visible_amount" type="int" />
+			<description>
+				Similar to [method get_next_visible_line_offset_from] however takes into account code wrap indexes. In the return vector, [code]x[/code] is the line, [code]y[/code] is the wrap index.
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:338]:
```suggestion
				Similar to [method get_next_visible_line_offset_from], but takes the code wrap indexes into account. In the returned vector, [code]x[/code] is the line, [code]y[/code] is the wrap index.
```

File: doc/classes/TextEdit.xml [hunk 21]
```diff
@@ -204,16 +316,67 @@
...
 			<return type="int" />
 			<argument index="0" name="tag" type="String" />
 			<description>
 				Returns OpenType feature [code]tag[/code].
 			</description>
 		</method>
+		<method name="get_saved_version" qualifiers="const">
+			<return type="int" />
+			<description>
+				Get the last tagged save version from [method tag_saved_version]
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:359]:
```suggestion
				Get the last tagged saved version from [method tag_saved_version]
```

File: doc/classes/TextEdit.xml [hunk 22]
```diff
@@ -265,6 +424,19 @@
...
+		<method name="get_total_visible_line_count" qualifiers="const">
+			<return type="int" />
+			<description>
+				Gets the total amount of lines that could be draw.
+			</description>
+		</method>
+		<method name="get_version" qualifiers="const">
+			<return type="int" />
+			<description>
+				Gets the current version of the [TextEdit].
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:439]: I think the "version" should be clarified, i.e. mention that it's related to undo history.

File: doc/classes/TextEdit.xml [hunk 23]
```diff
@@ -273,17 +445,44 @@
...
 			<return type="String" />
+			<argument index="0" name="position" type="Vector2" />
+			<description>
+				Gets the word at [code]position[/code].
+			</description>
+		</method>
+		<method name="get_word_under_caret" qualifiers="const">
+			<return type="String" />
+			<description>
+				Returns a [String] text with the word under the caret location.
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:458]:
```suggestion
				Returns a [String] text with the word under the caret's location.
```

File: doc/classes/TextEdit.xml [hunk 24]
```diff
@@ -273,17 +445,44 @@
...
+			<return type="bool" />
+			<description>
+				Returns if the user has IME text.
+			</description>
+		</method>
+		<method name="has_selection" qualifiers="const">
+			<return type="bool" />
 			<description>
-				Returns a [String] text with the word under the caret (text cursor) location.
+				Returns [code]true[/code] if the user is has a selection.
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:470]: Obvious mistake, should be reworded to make sense.

File: doc/classes/TextEdit.xml [hunk 25]
```diff
@@ -295,31 +494,43 @@
...
 			<return type="bool" />
 			<description>
+				Returns [code]true[/code] if the user is dragging their mouse for scrolling or selecting.
 			</description>
 		</method>
 		<method name="is_gutter_clickable" qualifiers="const">
 			<return type="bool" />
 			<argument index="0" name="gutter" type="int" />
 			<description>
+				Gets if the gutter is clickable.
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:504]:
```suggestion
				Returns whether the gutter is clickable.
```
Same in methods below.

File: doc/classes/TextEdit.xml [hunk 26]
```diff
@@ -346,12 +557,13 @@
...
 			<argument index="1" name="to_line" type="int" />
 			<description>
+				Merge the gutters from [code]from_line[/code] into [code]to_line[/code]. Only overwritable gutters will be copied.
 			</description>
 		</method>
 		<method name="paste">
 			<return type="void" />
 			<description>
-				Paste the current selection.
+				Paste the current selection. Can be overriden with [method _paste].
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:566]:
```suggestion
				Paste at the current selection. Can be overriden with [method _paste].
```
Or `into` idk

File: doc/classes/TextEdit.xml [hunk 27]
```diff
@@ -364,32 +576,44 @@
...
+		</method>
+		<method name="remove_text">
+			<return type="void" />
+			<argument index="0" name="from_line" type="int" />
+			<argument index="1" name="from_column" type="int" />
+			<argument index="2" name="to_line" type="int" />
+			<argument index="3" name="to_column" type="int" />
+			<description>
+				Removes text between the given positions.
+				[b]Note:[/b]This does not adjust the caret or selection as a result it can end up in an invalid position.
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:590]:
```suggestion
				[b]Note:[/b]This does not adjust the caret or selection, which as a result can end up in an invalid position.
```

File: doc/classes/TextEdit.xml [hunk 28]
```diff
@@ -364,32 +576,44 @@
...
+			<return type="Vector2i" />
+			<argument index="0" name="text" type="String" />
 			<argument index="1" name="flags" type="int" />
 			<argument index="2" name="from_line" type="int" />
-			<argument index="3" name="from_column" type="int" />
+			<argument index="3" name="from_colum" type="int" />
 			<description>
 				Perform a search inside the text. Search flags can be specified in the [enum SearchFlags] enum.
-				Returns an empty [code]Dictionary[/code] if no result was found. Otherwise, returns a [code]Dictionary[/code] containing [code]line[/code] and [code]column[/code] entries, e.g:
+				In the return vector, [code]x[/code] is the column, [code]y[/code] is the line. If no results are found both with equal [code]-1[/code].
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:601]:
```suggestion
				In the returned vector, [code]x[/code] is the column, [code]y[/code] is the line. If no results are found, both are equal to [code]-1[/code].
```

File: doc/classes/TextEdit.xml [hunk 29]
```diff
@@ -413,11 +637,39 @@
...
+				Selects the word under the caret.
+			</description>
+		</method>
+		<method name="set_caret_column">
+			<return type="void" />
+			<argument index="0" name="column" type="int" />
+			<argument index="1" name="adjust_viewport" type="bool" default="true" />
+			<description>
+				Moves the caret to the specified [code]column[/code] index.
+				If [code]adjust_viewport[/code] is set to [code]true[/code], the viewport will center at the caret position after the move occurs.
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:652]:
```suggestion
				If [code]adjust_viewport[/code] is [code]true[/code], the viewport will center at the caret position after the move occurs.
```

File: doc/classes/TextEdit.xml [hunk 30]
```diff
@@ -426,41 +678,47 @@
 			<argument index="1" name="object" type="Object" />
 			<argument index="2" name="callback" type="StringName" />
 			<description>
+				Set a custom draw method for the gutter. The callback method must take the following args: [code]line : int, gutter : int, Area : Rect2[/code]
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:681]:
```suggestion
				Set a custom draw method for the gutter. The callback method must take the following args: [code]line: int, gutter: int, Area: Rect2[/code].
```
Not sure about these spaces tbh.

File: doc/classes/TextEdit.xml [hunk 31]
```diff
@@ -426,41 +678,47 @@
...
 			</description>
 		</method>
 		<method name="set_gutter_draw">
 			<return type="void" />
 			<argument index="0" name="gutter" type="int" />
 			<argument index="1" name="draw" type="bool" />
 			<description>
+				Set whether the gutter should be drawn.
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:689]:
```suggestion
				Sets whether the gutter should be drawn.
```

File: doc/classes/TextEdit.xml [hunk 32]
```diff
@@ -527,12 +814,34 @@
 				Sets OpenType feature [code]tag[/code]. More info: [url=https://docs.microsoft.com/en-us/typography/opentype/spec/featuretags]OpenType feature tags[/url].
 			</description>
 		</method>
+		<method name="set_overtype_mode_enabled">
+			<return type="void" />
+			<argument index="0" name="enabled" type="bool" />
+			<description>
+				If [code]True[/code] set the user into overtype mode. When the user types it will override existing text.
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:821]:
```suggestion
				If [code]true[/code], sets the user into overtype mode. When the user types in this mode, it will override existing text.
```

File: doc/classes/TextEdit.xml [hunk 33]
```diff
@@ -527,12 +814,34 @@
...
+			</description>
+		</method>
+		<method name="set_search_flags">
+			<return type="void" />
+			<argument index="0" name="flags" type="int" />
+			<description>
+				Set the search flags. This is used with [method set_search_text] to highlight occurrences.
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:828]:
```suggestion
				Sets the search flags. This is used with [method set_search_text] to highlight occurrences.
```
Same below™

File: doc/classes/TextEdit.xml [hunk 34]
```diff
@@ -542,6 +851,29 @@
 				Sets the tab size for the [TextEdit] to use.
 			</description>
 		</method>
+		<method name="set_tooltip_request_func">
+			<return type="void" />
+			<argument index="0" name="object" type="Object" />
+			<argument index="1" name="callback" type="StringName" />
+			<argument index="2" name="data" type="Variant" />
+			<description>
+				Provide custom tooltip text. The callback method must take the following args: [code]hovered_word : String, data : Variant[/code]
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:860]:
```suggestion
				Provide custom tooltip text. The callback method must take the following args: [code]hovered_word: String, data: Variant[/code]
```

File: doc/classes/TextEdit.xml [hunk 35]
```diff
@@ -600,12 +934,18 @@
...
-			If [code]true[/code], read-only mode is enabled. Existing text cannot be modified and new text cannot be added.
-		</member>
 		<member name="scroll_horizontal" type="int" setter="set_h_scroll" getter="get_h_scroll" default="0">
 			If there is a horizontal scrollbar, this determines the current horizontal scroll value in pixels.
 		</member>
+		<member name="scroll_past_end_of_file" type="bool" setter="set_scroll_past_end_of_file_enabled" getter="is_scroll_past_end_of_file_enabled" default="false">
+			Allow scrolling past the last line into "virtual" space.
+		</member>
+		<member name="scroll_smooth" type="bool" setter="set_smooth_scroll_enable" getter="is_smooth_scroll_enabled" default="false">
+			Scrolling over the text rather then jumping.
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:944]: Sounds like something that should be reworded.

File: doc/classes/TextEdit.xml [hunk 36]
```diff
@@ -789,13 +1102,58 @@
...
+			Custom draw.
+		</constant>
 	</constants>
 	<theme_items>
 		<theme_item name="background_color" data_type="color" type="Color" default="Color(0, 0, 0, 0)">
 			Sets the background [Color] of this [TextEdit].
 		</theme_item>
 		<theme_item name="caret_background_color" data_type="color" type="Color" default="Color(0, 0, 0, 1)">
-			[Color] of the text behind the caret when block caret is enabled.
+			[Color] of the text behind the caret with a block caret.
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:1156]:
```suggestion
			[Color] of the text behind the caret when using a block caret.
```

File: scene/gui/code_edit.cpp [hunk 37]
```diff
@@ -544,58 +554,118 @@ Control::CursorShape CodeEdit::get_cursor_shape(const Point2 &p_pos) const {
...
+void CodeEdit::_handle_unicode_input(const uint32_t p_unicode) {
+	bool had_selection = has_selection();
 	if (had_selection) {
 		begin_complex_operation();
 		delete_selection();
 	}
 
-	// Remove the old character if in insert mode and no selection.
-	if (is_insert_mode() && !had_selection) {
+	/* Remove the old character if in insert mode and no selection. */
```
Comment (KoBeWi) [inline scene/gui/code_edit.cpp:567]: Why did you change the comment style? `//` is the preferred one.
Also it should say `overtype`.

File: scene/gui/code_edit.cpp [hunk 38]
```diff
@@ -544,58 +554,118 @@ Control::CursorShape CodeEdit::get_cursor_shape(const Point2 &p_pos) const {
...
+				remove_text(prev_line, prev_column, cl, cc);
+			}
+			set_caret_line(prev_line, false, true);
+			set_caret_column(prev_column);
+			return;
+		}
+	}
+
+	/* For space indentation we need to do a simple unindent if there are no chars to the left, acting in the */
+	/* same way as tabs.                                                                                      */
```
Comment (KoBeWi) [inline scene/gui/code_edit.cpp:655]:
```suggestion
	// For space indentation we need to do a simple unindent if there are no chars to the left, acting in the
	// same way as tabs.
```

File: doc/classes/TextEdit.xml [hunk 39]
```diff
@@ -364,32 +576,44 @@
...
+				if  result.x != -1:
 				    # Result found.
-				    var line_number = result.line
-				    var column_number = result.column
+				    var line_number = result.y
+				    var column_number = result.x
 				[/gdscript]
 				[csharp]
-				int[] result = Search("print", (uint)TextEdit.SearchFlags.WholeWords, 0, 0);
+				Vector2 result = Search("print", (uint)TextEdit.SearchFlags.WholeWords, 0, 0);
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:611]:
```suggestion
				Vector2i result = Search("print", (uint)TextEdit.SearchFlags.WholeWords, 0, 0);
```

**Comment by KoBeWi** [comment]
Lots of documentation mistakes, but the code looks ok at a glance.
I noticed that you remove `cache.` in lots of places. What was it and why is being removed?

**Comment by Paulb23** [comment]
`cache` was a `struct` that mainly held theme data. It wasn't really useful as a "cache" and contained some useless/duplicate properties. Removed it as it didn't really add anything apart from extra complexity and to clean up the properties.

File: doc/classes/TextEdit.xml [hunk 40]
```diff
@@ -108,46 +114,99 @@
 				Deselects the current selection.
 			</description>
 		</method>
+		<method name="end_complex_operation">
+			<return type="void" />
+			<description>
+				Ends a multipart edit, started with [method begin_complex_operation]. If called outside a complex operation, the current operation is push unto the undo/redo stack.
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:120]:
```suggestion
				Ends a multipart edit, started with [method begin_complex_operation]. If called outside a complex operation, the current operation is pushed unto the undo/redo stack.
```

File: doc/classes/TextEdit.xml (hunk 17 above)
Comment (KoBeWi) [reply doc/classes/TextEdit.xml]: Still missing `s`.

File: doc/classes/TextEdit.xml [hunk 41]
```diff
@@ -527,12 +814,34 @@
...
+			<argument index="0" name="enabled" type="bool" />
+			<description>
+				If [code]true[/code], sets the user into overtype mode. When the user types in this mode, it will override existing text.
+			</description>
+		</method>
+		<method name="set_search_flags">
+			<return type="void" />
+			<argument index="0" name="flags" type="int" />
+			<description>
+				Sets the search flags. This is used with [method set_search_text] to highlight occurrences. Arguments passed into [method search], should be passed here.
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:828]:
```suggestion
				Sets the search flags. This is used with [method set_search_text] to highlight occurrences. Arguments passed into [method search] should be passed here.
```
The last sentence about arguments sounds a bit unclear.

**Review by KoBeWi — APPROVED** [review]

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!
